libexec/mass-testing/test-harness

```diff
@@ -0,0 +1,235 @@
+        if pid_has_signal(process.pid, signal.SIGABRT):
+            main_pid = process.pid
+            break
+
+    if main_pid is None:
+        # Bail
+        root_process.send_signal(signal.SIGINT)
+        return
+
+    kill(main_pid, signal.SIGABRT)
+    try:
+        process.wait(30)
+    except psutil.TimeoutExpired:
+        kill(main_pid, signal.SIGKILL)
+        process.wait()
+    except psutil.ChildProcessError:
+        # Process died before we could `wait`
+        pass
+
+
+def patch_cmd(original_cmd: List[str], input_: str | None) -> List[str]:
+    result = []
+    for part in original_cmd:
+        if part == input_:
+            result.append("$INPUT")
+        elif Path(part).resolve() == (SCRIPT_DIR / "revng-test"):
+            result.append("revng")
+        else:
+            result.append(part)
+    return result
+
+
+def parse_args():
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--input", help="Input binary")
+    parser.add_argument("--output", required=True, help="Output directory")
+    parser.add_argument(
+        "--timeout", default=300, type=int, help="Time to timeout, in seconds (default: 300)"
+    )
+
+    args, rest = parser.parse_known_args()
+    if rest[0] == "--":
+        rest.pop(0)
+    return (args, rest)
+
+
+def main():
+    args, cmd = parse_args()
+    output_dir = Path.cwd() / args.output
+
+    time_file = NamedTemporaryFile()
+    time_cmd = [
+        "env",
+        "time",
+        "--quiet",
+        f"--format={_time_json_format}",
+        f"--output={time_file.name}",
+        "--",
+    ]
+
+    output_data = {
+        "command": patch_cmd(cmd, args.input),
+    }
+
+    base_env = os.environ.copy()
+
+    if "LLVM_SYMBOLIZER_PATH" in os.environ:
+        base_env["SYMBOLIZER_WRAPPER_SYMBOLIZER_PATH"] = base_env["LLVM_SYMBOLIZER_PATH"]
+    else:
+        symbolizer_path = which("llvm-symbolizer")
+        assert symbolizer_path is not None
+        base_env["SYMBOLIZER_WRAPPER_SYMBOLIZER_PATH"] = symbolizer_path
+
+    base_env["SYMBOLIZER_WRAPPER_OUTPUT"] = str(output_dir / "stacktrace.json")
+    base_env["LLVM_SYMBOLIZER_PATH"] = str((SCRIPT_DIR / "symbolizer_wrapper").resolve())
+    base_env["TEST_OUTPUT_DIR"] = args.output
+
+    if args.input is not None:
+        os.symlink(os.path.relpath(args.input, output_dir), output_dir / "input")
+        base_env["TEST_INPUT"] = args.input
+
+    # Setup as child subreaper, this avoids processes "escaping"
+    # the test harness
+    set_child_subreaper()
+
+    with open(output_dir / "output.log", "wb") as log:
+        process = Popen(
+            [*time_cmd, *cmd],
+            stdout=log,
+            stderr=STDOUT,
+            env=base_env,
+        )
+
+    try:
+        process.wait(args.timeout)
+        timed_out = False
+    except TimeoutExpired:
+        handle_timeout(process)
+        timed_out = True
+
+    # Wait for things to settle and then take care of children
+    sleep(10)
```

Comment:
Do we really really really need this?
10 seconds is a long time.

Reply:
Eh, in general the sleep there is to account for `time` & the other wrappers to gracefully terminate. I can cut it down to say 5 seconds, but the reality is that it's a random number that "looks good".

Reply:
Can't at least check every second if the root process is terminated *up to* 10 seconds?

Reply:
That sounds sensible, I'll do it that way.